### ISSUE TITLE
Bug: [prefer-nullish-coalescing] Should not report on false | undefined values by default

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#11092]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.8.2&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gWygTwCMQAueAcijPgB9yCyBuAKFElgWXSwDMoIICUMAGtSAbTJgyAXWatoceNwCuKMBgCWqRHCgYQANT7KQaABQBKUphgaUAczHSa8VaG52QwZkxVrN2sjYAA7K%2BgDKGLYO5lbwNnaOzgDeTPCIqJjwdvow2F4aeggAvDj4RPDFVeSU8ABkdTogRUYQJrE%2B6ZxZAG7GJdkoufnAhfo0tLz8giKd8HAYyjAo8H1tIMwAvkxAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6RAM0WlqYSNkAC1pkA9gEMkyMswDm6KL2jjokcGAC%2BILUA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
declare const maybe: 'a' | 'b';
declare const fallback: ['c'];
declare function createValues(): string[] | undefined;

function computeStrings(): string[] {
  const intermediate = maybe === 'a' && createValues();

  const value = intermediate || fallback;

  return value;
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/prefer-nullish-coalescing": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

`intermediate` is `false | string[] | undefined` and its parent function returns `string[]`. The `||` is the right operator to use to make `value` a `string[]`, not a `string[] | false`.

### Actual Result

```plaintext
Prefer using nullish coalescing operator (`??`) instead of a logical or (`||`), as it is a safer operator. 8:30 - 8:32
```

### Additional Info

Related: #7055, #10356, and probably others, searching gives a lot of false positive results...

#10356 notes that this gets suggested a good amount. I personally was surprised to re-discover that the default behavior of the rule is to report in a way that is sometimes incorrect/unnecessary.

Proposal: I think we should switch the rule to ignore `false | undefined | (always-truthy)`s by default. 

💖

**Comment by kirkwaiblinger** [comment]
I'd probably prefer `ignorePrimitives: true` being default and using strict-boolean-expressions to handle the primitive cases.
